### PR TITLE
GH-141 - Handle targetIsCollection flag in CompoundQueryCondition

### PR DESCRIPTION
**Pull request by ractive**
Keep the targetIsCollection in the CompoundQueryCondition so that the element serializer is used to serialize the query and not the collection serializer.

See GH-141

### CODE REVIEW

File: src/test/java/org/mongojack/TestQuerySerialization.java [hunk 1]
```diff
@@ -118,7 +137,10 @@ public void testAll() {
         coll.save(o);
 
         // Ensure that the serializer actually worked
+        // with DBCursor
         assertThat(coll.find().all("items", o1).toArray(), hasSize(1));
+        // with DBQuery
+        assertThat(coll.find(DBQuery.all("items", o1)).toArray(), hasSize(1));
```
Comment (ractive) [inline src/test/java/org/mongojack/TestQuerySerialization.java]: This test actually failed.

File: src/test/java/org/mongojack/TestQuerySerialization.java [hunk 2]
```diff
@@ -88,7 +96,10 @@ public void testLessThan() {
         // Ensure that the serializer actually worked
         assertThat((Integer) coll.getDbCollection().findOne().get("i"),
                 equalTo(15));
+        // with DBCursor
         assertThat(coll.find().lessThan("i", 12).toArray(), hasSize(1));
+        // with DBQuery
+        assertThat(coll.find(DBQuery.lessThan("i", 12)).toArray(), hasSize(1));
```
Comment (ractive) [inline src/test/java/org/mongojack/TestQuerySerialization.java]: Added DBQuery versions for all tests here.

File: src/main/java/org/mongojack/internal/util/SerializationUtils.java [hunk 3]
```diff
@@ -196,13 +195,17 @@ private static Object serializeQueryCondition(
         } else {
             CompoundQueryCondition compound = (CompoundQueryCondition) condition;
             if (!key.startsWith("$")) {
```
Comment (benmccann) [inline src/main/java/org/mongojack/internal/util/SerializationUtils.java]: can you change this to use `isOperator` as well

File: src/test/java/org/mongojack/TestQuerySerialization.java [hunk 4]
```diff
@@ -56,17 +56,25 @@ public void setUp() {
...
                 coll.find()
                         .in("_id", id, new org.bson.types.ObjectId().toString())
                         .toArray(), hasSize(1));
+        // with DBQuery
+        assertThat(
+                coll.find(DBQuery.in("_id", id, new org.bson.types.ObjectId().toString()))
+                        .toArray(), hasSize(1));
     }
 
     @Test @Ignore // This test needs to be fixed
```
Comment (benmccann) [inline src/test/java/org/mongojack/TestQuerySerialization.java]: can this `@Ignore` be removed? in https://github.com/mongojack/mongojack/pull/144 you added a comment that this PR fixes it

File: src/main/java/org/mongojack/internal/util/SerializationUtils.java (hunk 3 above)
Comment (ractive) [reply src/main/java/org/mongojack/internal/util/SerializationUtils.java]: Done

File: src/test/java/org/mongojack/TestQuerySerialization.java (hunk 4 above)
Comment (ractive) [reply src/test/java/org/mongojack/TestQuerySerialization.java]: Good catch! :-)
This probably survived when merging.

**Comment by ractive** [comment]
I'm pretty confident that this could be merged now. WDYT?

**Comment by benmccann** [comment]
I'm no longer maintaining this repo. @cfogrady would you like to review this PR?

**Comment by cfogrady** [comment]
I'll check it out before this weekend.

File: src/main/java/org/mongojack/internal/util/SerializationUtils.java [hunk 5]
```diff
@@ -195,14 +194,18 @@ private static Object serializeQueryCondition(
             return serializedConditions;
         } else {
             CompoundQueryCondition compound = (CompoundQueryCondition) condition;
-            if (!key.startsWith("$")) {
-                serializer = findQuerySerializer(false, key, serializerProvider, serializer);
```
Comment (cfogrady) [inline src/main/java/org/mongojack/internal/util/SerializationUtils.java]: Can you add this to the DocumentSerializationUtils as well? I think ideally the portion of the library dealing with Mongo 2.X API should be deprecated.

Comment (ractive) [reply src/main/java/org/mongojack/internal/util/SerializationUtils.java]: Done

**Pull request merged**